source/cfa.h

```diff
@@ -0,0 +1,133 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SPVTOOLS_CFA_H_
+#define SPVTOOLS_CFA_H
+
+#include <algorithm>
+#include <cassert>
+#include <functional>
+#include <map>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+using std::find;
+using std::function;
+using std::get;
+using std::pair;
+using std::unordered_map;
+using std::unordered_set;
+using std::vector;
+
+namespace spvtools {
+
+template<class BB> class CFA {
+  using bb_ptr = BB*;
+  using cbb_ptr = const BB*;
+  using bb_iter = typename std::vector<BB*>::const_iterator;
+  using get_blocks_func =
+    std::function<const std::vector<BB*>*(const BB*)>;
+
+  struct block_info {
+    cbb_ptr block;  ///< pointer to the block
+    bb_iter iter;   ///< Iterator to the current child node being processed
+  };
+
+  /// Returns true if a block with @p id is found in the @p work_list vector
+  ///
+  /// @param[in] work_list  Set of blocks visited in the the depth first traversal
+  ///                       of the CFG
+  /// @param[in] id         The ID of the block being checked
+  ///
+  /// @return true if the edge work_list.back().block->id() => id is a back-edge
+  bool FindInWorkList(const std::vector<block_info>& work_list, uint32_t id);
+
+public:
+  /// @brief Depth first traversal starting from the \p entry BasicBlock
+  ///
+  /// This function performs a depth first traversal from the \p entry
+  /// BasicBlock and calls the pre/postorder functions when it needs to process
+  /// the node in pre order, post order. It also calls the backedge function
+  /// when a back edge is encountered.
+  ///
+  /// @param[in] entry      The root BasicBlock of a CFG
+  /// @param[in] successor_func  A function which will return a pointer to the
+  ///                            successor nodes
+  /// @param[in] preorder   A function that will be called for every block in a
+  ///                       CFG following preorder traversal semantics
+  /// @param[in] postorder  A function that will be called for every block in a
+  ///                       CFG following postorder traversal semantics
+  /// @param[in] backedge   A function that will be called when a backedge is
+  ///                       encountered during a traversal
+  /// NOTE: The @p successor_func and predecessor_func each return a pointer to a
+  /// collection such that iterators to that collection remain valid for the
+  /// lifetime of the algorithm.
+  void DepthFirstTraversal(const BB* entry,
```

Comment:
Question (not a request for change):  Could this have been a static method instead, and still as easily used?

This method doesn't require any external state, so I think it could have been a static method.   I *think* arg-dependent-lookup would have made it usable as a static method, and so we wouldn't need to create the object.